xcube/core/store/zarrdescriber.py

```diff
@@ -0,0 +1,183 @@
+                                dataset_dims[var_dim] = var_shape[i]
+                data_vars[variable_name] = dict(name=variable_name,
+                                                dtype=zarr_attrs.get('data_type', folder_dtype),
+                                                dims=var_dims,
+                                                attrs=zarr_attrs)
+
+        # Remove data_vars that are actually dimensions
+        for dim in dataset_dims:
+            if dim in data_vars:
+                data_vars.pop(dim)
+        to_be_removed = []
+        for data_var in data_vars.keys():
+            if data_var.endswith('bnds') or data_var.endswith('bounds'):
+                to_be_removed.append(data_var)
+                continue
+            if known_variable_names is not None and data_var not in known_variable_names:
+                to_be_removed.append(data_var)
+        for remove in to_be_removed:
+            data_vars.pop(remove)
+
+        descriptor_dict['dims'] = dataset_dims
+        descriptor_dict['data_vars'] = data_vars
+        return DatasetDescriptor.from_dict(descriptor_dict)
+
+    @staticmethod
+    def _set_if_included(source: dict,
+                         target: dict,
+                         source_attrs: Sequence[str],
+                         target_attr: str):
+        for a in source_attrs:
+            if a not in source:
+                return
+        if len(source_attrs) == 1:
+            target[target_attr] = source[source_attrs[0]]
+        else:
+            target[target_attr] = tuple([source[attr_name] for attr_name in source_attrs])
+
+    @abstractmethod
+    def _open_file(self, data_id: str, file: str):
+        pass
+
+    @abstractmethod
+    def _list_zarr_dir_content(self, data_id: str):
+        pass
+
+
+class S3ZarrDescriber(ZarrDescriber):
+    """
+    A describer that derives a DatasetDescriptor from a S3 file system.
+
+    :param s3: A S3 File System
+    :param bucket_name: The name of the bucket within the file system in
+    which the zarr data is located.
+    """
+
+    def __init__(self, s3: s3fs.S3FileSystem, bucket_name: str):
+        self._s3 = s3
+        self._bucket_name = bucket_name
+
+    def _open_file(self, data_id: str, file: str):
+        return self._s3.open(f'{self._bucket_name}/{data_id}/{file}', 'r')
+
+    def _list_zarr_dir_content(self, data_id: str):
+        return self._s3.ls(f'{self._bucket_name}/{data_id}')
+
+
+class DirectoryZarrDescriber(ZarrDescriber):
+    """
+    A describer that derives a DatasetDescriptor from a file system.
+
+    :param path: The base path within the file system in
+    which the zarr data is located.
+    """
+
+    def __init__(self, path: str):
+        self._path = path
+
+    def _open_file(self, data_id: str, file: str):
+        return open(f'{self._path}/{data_id}/{file}', 'r')
+
+    def _list_zarr_dir_content(self, data_id: str):
+        return os.listdir(f'{self._path}/{data_id}')
```

Comment:
Move this into a separate implementation package, e.g. into `accessors` or a new one. As an implementation detail, it does not belong to the data store framework.

For the future, I'd prefer, if we model a data _registry_ or data _index_ instead, so 

- we are no longer bound to files (_open_file, _list_zarr_dir_content);
- the registry may be backed by an efficient cloud indexing software, e.g. Apache Solr, MongoDB, REDIS, ...

Reply:
Do you mean the whole module? I'd agree. I think a new package 'describers' would be appropriate. Okay?

Reply:
No, this is an implementation detail that we want to hide. It does not deserve an own package. Move this module as a private one into `accessors` or `stores` and don't let its classes escape into the public store API.